File: app-bar-provider/src/main/java/com/sequenia/app_bar_provider/AppBarProviderImp.java
package com.sequenia.app_bar_provider;

import android.content.Context;
import android.content.res.Resources;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;

import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.appbar.CollapsingToolbarLayout;


/**
 * Created by Ringo on 06.02.2018.
 * <p>
 * Методы по работе с AppBar
 * - добавление custom view в @{@link Toolbar}
 * - видимость кнопки home в @{@link Toolbar}
 * - изменение иконки home в @{@link Toolbar}
 * - добаваление view в @{@link CollapsingToolbarLayout}
 * - добавление view в AppBar
 * - создание view с родителем AppBar
 */
public class AppBarProviderImp {

    private AppBarLayout appBar;
    private Toolbar toolbar;
    private ViewGroup collapsingContent;
    private CollapsingToolbarLayout collapsingToolbarLayout;

    public AppBarProviderImp(AppBarViews appBarViews) {
        appBar = appBarViews.getAppBar();
        toolbar = appBarViews.getToolbar();
        collapsingToolbarLayout = appBarViews.getCollapsingToolbarLayout();
        collapsingContent = appBarViews.getCollapsingContent();
        toolbar.setContentInsetStartWithNavigation(0);
    }

    void addViewToCollapsingView(View view) {
        collapsingContent.addView(view, collapsingContent.getChildCount());
    }

    void removeViewFromCollapsingView(View view) {
        collapsingContent.removeView(view);
    }

    View setCustomToolbarView(int layoutRes, ActionBar actionBar, LayoutInflater inflater) {
        if (actionBar == null) {
            return null;
        }

        if (layoutRes == -1) {
            actionBar.setCustomView(null);
        } else {
            ActionBar.LayoutParams params = new ActionBar.LayoutParams(
                    ActionBar.LayoutParams.MATCH_PARENT,
                    ActionBar.LayoutParams.MATCH_PARENT
            );

            View customView = inflater.inflate(layoutRes, null);
            actionBar.setCustomView(customView, params);
        }

        return actionBar.getCustomView();
    }

    void setBackButtonVisibility(boolean visibility, ActionBar actionBar) {
        if (actionBar == null) {
            return;
        }

        int displayOptions = ActionBar.DISPLAY_SHOW_CUSTOM |
                (visibility ? ActionBar.DISPLAY_HOME_AS_UP : 0);
        actionBar.setDisplayOptions(displayOptions);
    }

    void setNeedScrollAppBar(boolean needScroll, int flags) {
        if (toolbar != null && collapsingToolbarLayout != null) {
            AppBarLayout.LayoutParams params =
                    (AppBarLayout.LayoutParams) collapsingToolbarLayout.getLayoutParams();
            params.setScrollFlags(!needScroll ?
                    0 : AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL | flags
            );
            collapsingToolbarLayout.setLayoutParams(params);
        }
    }

    void setToolbarVisibility(boolean visibility) {
        toolbar.setVisibility(visibility ? View.VISIBLE : View.GONE);
    }

    void setAppBarVisibility(boolean visibility) {
        appBar.setVisibility(visibility ? View.VISIBLE : View.GONE);
    }

    View inflateViewForAppBar(int layoutRes, LayoutInflater inflater) {
        return inflater.inflate(layoutRes, appBar, false);
    }

    void addViewToAppBar(View view) {
        appBar.addView(view, appBar.getChildCount());
    }

    void removeViewFromAppBar(View view) {
        appBar.removeView(view);
    }

    void setHomeAsUpIndicator(ActionBar actionBar, Integer drawableRes) {
        if (actionBar == null || appBar == null) {
            return;
        }

        if (drawableRes == null) {
            drawableRes = getDefaultHomeAsUpIndicator();
        }

        // по каким-то причинам не удалось достать иконку по умолчанию
        if (drawableRes == null) {
            return;
        }

        actionBar.setHomeAsUpIndicator(drawableRes);
    }

    private Integer getDefaultHomeAsUpIndicator() {
        Context context = appBar.getContext();

        if (context == null) {
            return null;
        }

        Resources.Theme theme = context.getTheme();

        if (theme == null) {
            return null;
        }

        TypedValue typedValue = new TypedValue();
        theme.resolveAttribute(android.R.attr.homeAsUpIndicator, typedValue, true);
        return typedValue.resourceId;
    }
}

File: app-bar-provider/src/main/java/com/sequenia/app_bar_provider/AppBarProviderImp.kt
package com.sequenia.app_bar_provider

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.ActionBar
import androidx.appcompat.widget.Toolbar
import androidx.core.view.isVisible
import androidx.core.view.updateLayoutParams
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.CollapsingToolbarLayout

open class AppBarProviderImp(
    appBarViews: AppBarViews,
) {
    private val appBar: AppBarLayout?
    private var toolbar: Toolbar?
    private var collapsingContent: ViewGroup? = null
    private var collapsingToolbarLayout: CollapsingToolbarLayout? = null

    init {
        appBar = appBarViews.getAppBar()
        toolbar = appBarViews.getToolbar()
        collapsingToolbarLayout = appBarViews.getCollapsingToolbarLayout()
        collapsingContent = appBarViews.getCollapsingContent()
        toolbar?.contentInsetStartWithNavigation = 0
    }

    fun addViewToCollapsingView(view: View) {
        collapsingContent?.let { it.addView(view, it.childCount) }
    }

    fun removeViewFromCollapsingView(view: View?) {
        collapsingContent?.removeView(view)
    }

    fun setCustomToolbarView(
        layoutRes: Int,
        actionBar: ActionBar,
        inflater: LayoutInflater,
    ): View? {
        if (layoutRes == -1) {
            actionBar.customView = null
            return null
        }

        val params = ActionBar.LayoutParams(
            ActionBar.LayoutParams.MATCH_PARENT,
            ActionBar.LayoutParams.MATCH_PARENT
        )

        val customView = inflater.inflate(layoutRes, null)
        actionBar.setCustomView(customView, params)

        return actionBar.customView
    }

    fun setBackButtonVisibility(visibility: Boolean, actionBar: ActionBar) {
        val displayOptions = ActionBar.DISPLAY_SHOW_CUSTOM or
                if (visibility) ActionBar.DISPLAY_HOME_AS_UP else 0
        actionBar.displayOptions = displayOptions
    }

    fun setNeedScrollAppBar(needScroll: Boolean, flags: Int) {
        collapsingToolbarLayout?.updateLayoutParams<AppBarLayout.LayoutParams> {
            this.scrollFlags =
                if (!needScroll) 0 else AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL or flags
        }
    }

    fun setToolbarVisibility(isVisible: Boolean) {
        toolbar?.isVisible = isVisible
    }

    fun setAppBarVisibility(isVisible: Boolean) {
        appBar?.isVisible = isVisible
    }

    fun inflateViewForAppBar(layoutRes: Int, inflater: LayoutInflater): View? {
        return inflater.inflate(layoutRes, appBar, false)
    }

    fun addViewToAppBar(view: View) {
        appBar?.let { it.addView(view, it.childCount) }
    }

    fun removeViewFromAppBar(view: View) {
        appBar?.removeView(view)
    }

    fun setHomeAsUpIndicator(actionBar: ActionBar, drawableRes: Int) {
        actionBar.setHomeAsUpIndicator(drawableRes)
    }

    fun setToolbarTitle(title: String) {
        toolbar?.title = title
    }
}